content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+Requests executed: 50000, memory allocated: 31547 bytes/request
+Requests executed: 60000, memory allocated: 31545 bytes/request
+Requests executed: 70000, memory allocated: 31537 bytes/request
+Requests executed: 80000, memory allocated: 31624 bytes/request
+Requests executed: 90000, memory allocated: 31703 bytes/request
+Requests executed: 100000, memory allocated: 31682 bytes/request
+----
+
+As we see, there's some warm-up phase during allocation rates still get down,
+but after ~20 K requests, the allocation per request is fairly stable,
+with a volatility of ~1% when averaged out over 10K requests.
+I.e. this initial phase should be excluded during the actual test.
+
+To emphasize the key part again, this allocation is per _request_, it is independent from wall clock time and thus is not dependent on the machine running the test (i.e. the test should behave the same when running on a developer laptop and on a CI machine), nor is it subject to volatility induced by other workloads running concurrently.
+
+Based on that, the actual test could look like so:
+
+[source, java]
+----
+@Test
+@EnableEvent("jdk.ObjectAllocationInNewTLAB")
+@EnableEvent("jdk.ObjectAllocationOutsideTLAB")
+public void retrieveTodo() throws Exception {
+  Random r = new Random();
+  HttpClient client = HttpClient.newBuilder().build();
+
+  for (int i = 1; i<= 20_000; i++) { // <1>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+  jfrEvents.reset();
+
+  for (int i = 1; i<= 10_000; i++) { // <2>
+    executeRequest(r, client);
+  }
+
+  jfrEvents.awaitEvents();
+
+  long sum = jfrEvents.filter(this::isObjectAllocationEvent)
+      .filter(this::isRelevantThread)
+      .mapToLong(this::getAllocationSize)
+      .sum();
+
+  assertThat(sum / 10_000).isLessThan(33_000); // <3>
+}
+----
+<1> Warm-up phase
+<2> The actual test phase
+<3> Assert the memory allocation per request is within the expected boundary
+
+Now let's assume we've wrapped up the initial round of work on this application, and its tests have been passing on CI for a while.
+One day, the `retrieveTodo()` performance test method fails though:
+
+[source]
+----
+java.lang.AssertionError: 
+Expecting:
+ <388370L>
+to be less than:
+ <33000L> 
+----
+
+Wow, it's suddenly allocating more than 10 times more memory per request than before!
+What has happened?
+To find the answer, we can take a look at the test's JFR recording, which JfrUnit persists under _target/jfrunit_:
+
+[source,bash]
+----
+ls target/jfrunit
+
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-createTodo.jfr
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-retrieveTodo.jfr
+----
+
+Let's open the *.jfr file for the failing test in JDK Mission Control (JMC) in order to analyse all the recorded events
+(note that the recording will always contain some JfrUnit-internal events which are needed for synchronizing the recording stream and the events exposed to the test).
```

Comment:
would it make sense and be possible to exclude these internal events from the recording? if not during, maybe as a post-processing step or is it simply not worth it?

Reply:
It cannot be excluded, as it's required as sort of a synchronization barrier for making sure all previously produced events have been received (this all is async). I suppose you could post-process the recording, but I don't think it's worth it actually.